docs/flip/1777-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+# FLIP: Default dtypes
+
+
+- Start Date: 2022-01-11
+- FLIP PR: [#1776](https://github.com/google/flax/pull/1776)
+- FLIP Issue: [#1777](https://github.com/google/flax/issues/1777)
+
+
+## Summary
+
+This FLIP proposes to replace the default dtype which is currently hard-coded to float32, and instead use the JAX type promotion results to derive a default dtype from the input and parameters of a layer.
+
+
+## Motivation
+
+Currently, linen Modules always produce float32 outputs regardless of input and parameter dtypes. Half-precision types like float16 and bfloat16 are supported by explicitly passing the half-precision type to each layer. The way this is currently implemented is that each layer has a dtype argument with float32 as the default value. The layer guarantees that this dtype will be the return type of the result returned by `__call__`.
+
+The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.
```

Comment:
It feels like that's never a good idea. I'm wondering if we should always raise an error if one of the `xs` tree leaves is complex but `dtype` is not. Users should maybe remove imaginary part by themselves if that's really what they want to do.
(Maybe it's a contrived example, but I can imagine cases where layers have their `dtype` set by parent modules based on assumptions without complex numbers in mind)

Reply:
Andreas: FYI I believe this entire FLIP was motivated mainly for complex numbers, and what you are describing is indeed the main issue that users using complex numbers ran into.

Reply:
@andsteing One of the big issues we face when using Linen with NetKet/Quantum Mechanics is that we usually want to specify the dtype of the parameters in the module (real/complex) but we want the output type to be decided by type promotion.
This is because often our input type can change and both be real/complex depending on what quantities we are computing, but the model itself does not change.

To put it differently, 95% of the time, in our use-cases, we treat layers such as dense as if they are 'matrices' and we are doing `Dense @ input`: we do not want to downcast types, and `Dense.dtype` for us matches the dtype of the internal data (aka, the parameters). 
Of course this is not consistent with flax of having modules be callables, where users do `Dense(input)` and where, now, `Dense.dtype` should reflect the output dtype.

This is why @jheek proposed to have a default `Dense.dtype = None`, which will follow default numpy type promotion, and if you want to override that the user has to specify it explicitly.

Reply:
I agree `Dense.dtype = None` would work as expected.

But I was wondering: Is there *ever* a case where `Dense.dtype = float32` would make sense with a complex input (or parameters)? If not, then we could throw an error in that case (for extra safety).

Reply:
Theoretically I'd like if that could be interpreted as single-precision and mean that complex128 would be casted to complex64.

However this rule is not used anywhere in Numpy (unfortunately) and therefore as a stop gap I think it's better to error.

Regardless, erroring now gives us the possibility of changing this behavior in the future.

Reply:
@andsteing we could indeed manually throw an error but I left it out of this FLIP (it could easily be added in a separate PR). The reasoning is that we would be more strict than NumPy in this case. And this silent bug is better solved at the JAX level where it's easy to catch and would apply more generally (will add to discussion).